### PR TITLE
Fix for low accuracy in Tensorflow training

### PR DESCRIPTION
**Pull request by pribalta**
Indexed_file_loader.h performs a sharded read of the index files supplied alongside the tfrecords for training. Thus, ReadSample() is called from multiple threads with different shard_id_. When all indexes have been read, we proceed to reset current_index_ to start again at the beginning of the shard.

This commit fixes a defect in the detection that the shard has arrived to the end, by comparing the value of current_index to the size of all the indexes. The fix consists in comparing against the end of the shard instead of the end of the indices.

Before the fix, the code would work correctly for 1 GPU when there is only 1 shard, but with >1 GPUs, it would result in a loss in accuracy due to---most likely---redundant reads.

Signed-off-by: Pablo Ribalta Lorenzo <pribalta@nvidia.com>

### CODE REVIEW

File: dali/pipeline/operators/reader/loader/indexed_file_loader.h [hunk 1]
```diff
@@ -94,6 +94,9 @@ class IndexedFileLoader : public Loader<CPUBackend> {
     ReadIndexFile(index_uris);
     size_t num_indices = indices_.size();
     current_index_ = num_indices/num_shards_ * shard_id_;
+    max_index_ = (shard_id_ != num_shards_ - 1)
+      ? num_indices/num_shards_ * (shard_id_ + 1)
```
Comment (JanuszL) [inline dali/pipeline/operators/reader/loader/indexed_file_loader.h]: How about just:
```
max_index_ = num_indices * (shard_id_ + 1) / num_shards_ 
```

Comment (pribalta) [reply dali/pipeline/operators/reader/loader/indexed_file_loader.h]: accepted

**Comment by pribalta** [comment]
Waiting till another run finishes and reports good results and I'll be merging this guy

**Comment by ptrendx** [comment]
I agree that it is a bug (good job finding it :-) ) but I do not agree fully with the fix, because it is not consistent with the rest of readers. You should instead change the reset function to set the current index to 0

**Pull request merged**